notebook-controller/tox.ini

```diff
@@ -0,0 +1,74 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+[tox]
+skipsdist = True
+skip_missing_interpreters = True
+
+[testenv]
+setenv =
+    PYTHONPATH={toxinidir}
+    PYTHONBREAKPOINT=ipdb.set_trace
+    CHARM_REPO=https://github.com/canonical/notebook-operators.git
+    CHARM_BRANCH=main
+    LOCAL_CHARM_DIR=charm_repo
+
+[testenv:unit]
+passenv = *
+allowlist_externals =
+    bash
+    tox
+    rockcraft
+deps =
+    charmed_kubeflow_chisme
+    juju~=2.9.0
+    pytest
+    pytest-operator
+    ops
+commands =
+    # build and pack rock
+    rockcraft clean
+    rockcraft pack
+    bash -c 'NAME=$(yq eval .name rockcraft.yaml) && \
+             VERSION=$(yq eval .version rockcraft.yaml) && \
+             ARCH=$(yq eval ".platforms | keys" rockcraft.yaml | awk -F " " '\''{ print $2 }'\'') && \
+             ROCK="$\{NAME\}_$\{VERSION\}_$\{ARCH\}" && \
+             sudo skopeo --insecure-policy copy oci-archive:$ROCK.rock docker-daemon:$ROCK:$VERSION && \
+             docker save $ROCK > $ROCK.tar'
+    # run rock tests
+    pytest -v --tb native --show-capture=all --log-cli-level=INFO {posargs} {toxinidir}/tests
+
+[testenv:integration]
```

Comment:
Are we going to start running integration tests from other repos for all the rocks?

Reply:
If I understood your question correctly, then Yes: For every ROCK that is managed by charm we will checkout that charms repo, update metadata with local ROCK images (local ROCK images is copied into microk8s cache), and run normal charm's integration tests.
If ROCK is not managed by charm, there will be unit tests that are executed. Those will differe from ROCK to ROCK and will be located in ROCK repository. In most cases they will act as smoke tests, i.e. check if all artifacts are present, that ROCK starts.
In some cases, ROCK might not start without extensive configuration and/or K8S resources, then the only testing we will do is integration tests.

Reply:
> Are we going to start running integration tests from other repos for all the rocks?

I.e for Pipeline ROCKs we should do the same. However, it all depends on the ROCK. Charm managed are easier to test using our integration test. Workload managed ROCKs are trickier and we might have unit (integrity/smoke) tests only for those.